MoreLinq/PendNode.cs

```diff
@@ -91,7 +91,9 @@ public IEnumerator<T> GetEnumerator()
                                 case 1: concat2 = item.Value; break;
                                 case 2: concat3 = item.Value; break;
                                 case 3: concat4 = item.Value; break;
+#pragma warning disable CA2201 // Do not raise reserved exception types
                                 default: throw new IndexOutOfRangeException();
```

Comment:
`ArgumentOutOfRangeException`

Reply:
This case cannot occur due to a caller-supplied argument so `ArgumentOutOfRangeException` doesn't seem to be the right call here. In fact, arriving here would constitute a bug in the implementation. Ideally, we could have used [`UnreachableException`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.unreachableexception), but that just got added in .NET 7.

Reply:
`NotSupportedException` with a message of "Bug in MoreLinq implementation." (needs better wording, but you get the idea) would probably be a better expression here then. `IndexOutOfRange` looks out of place and doesn't explain the intent of this `throw`. 

I'm good with it being done on another PR though.

Reply:
I've added a private version of `UnreachableException` with ac34b3055ea18b8e83435de2238716fa1521b9bf and thrown that here instead.